src/mixins/java/org/spongepowered/common/mixin/core/server/players/PlayerListMixin.java

```diff
@@ -521,56 +509,56 @@ public abstract class PlayerListMixin implements PlayerListBridge {
+                destinationPosition, originalWorld, originalPosition, destinationWorld, this.impl$originalRespawnDestination, destinationPosition,
+                originalPlayer, recreatedPlayer, this.impl$isGameMechanicRespawn, !keepAllPlayerData);
         SpongeCommon.post(event);
-        recreatedPlayer.setPos(event.destinationPosition().x(), event.destinationPosition().y(), event.destinationPosition().z());
+
         this.impl$isGameMechanicRespawn = false;
-        this.impl$originalDestination = null;
-        this.impl$newDestination = null;
-
-        final ServerLevel targetWorld = (ServerLevel) event.destinationWorld();
-        ((ServerPlayerBridge) recreatedPlayer).bridge$sendChangeDimension(
-            targetWorld.dimensionType(),
-            ((ClientboundRespawnPacketAccessor) packetIn).accessor$dimension(),
-            ((ClientboundRespawnPacketAccessor) packetIn).accessor$seed(),
-            recreatedPlayer.gameMode.getGameModeForPlayer(),
-            recreatedPlayer.gameMode.getPreviousGameModeForPlayer(),
-            targetWorld.isDebug(),
-            targetWorld.isFlat(),
-            keepAllPlayerData
-        );
+        newPlayer.setPos(event.destinationPosition().x(), event.destinationPosition().y(), event.destinationPosition().z());
+
+        return newPlayer.getX();
     }
 
     @Inject(method = "respawn", at = @At("RETURN"))
     private void impl$callRespawnPlayerPostEvent(final net.minecraft.server.level.ServerPlayer player, final boolean keepAllPlayerData, final CallbackInfoReturnable<net.minecraft.server.level.ServerPlayer> cir) {
-        final org.spongepowered.api.world.server.ServerWorld originalWorld = (org.spongepowered.api.world.server.ServerWorld) player.level;
-        final org.spongepowered.api.world.server.ServerWorld originalDestinationWorld = (org.spongepowered.api.world.server.ServerWorld) this.server.getLevel(this.impl$originalDestination == null ? Level.OVERWORLD : this.impl$originalDestination);
-        final org.spongepowered.api.world.server.ServerWorld destinationWorld = (org.spongepowered.api.world.server.ServerWorld) this.server.getLevel(this.impl$newDestination == null ? Level.OVERWORLD : this.impl$newDestination);
+        final ServerPlayer recreatedPlayer = (ServerPlayer) cir.getReturnValue();
+        final ServerWorld originalWorld = (ServerWorld) player.level;
 
-        final RespawnPlayerEvent.Post event = SpongeEventFactory.createRespawnPlayerEventPost(PhaseTracker.getCauseStackManager().currentCause(), destinationWorld, originalWorld, originalDestinationWorld, (ServerPlayer) cir.getReturnValue());
+        final RespawnPlayerEvent.Post event = SpongeEventFactory.createRespawnPlayerEventPost(PhaseTracker.getCauseStackManager().currentCause(),
+                recreatedPlayer.world(), originalWorld, this.impl$originalRespawnDestination, recreatedPlayer);
         SpongeCommon.post(event);
+
+        this.impl$originalRespawnDestination = null;
```

Comment:
I'm not fond of having this being the only reset for the respawn destination in the event of exceptions during respawns, I could see the field being leaked until the next player respawns, which means a world is leaked (unless the supplier is fixed).